plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,148 @@
+import io.airlift.log.Logger;
+import io.trino.collect.cache.EvictableCacheBuilder;
+import io.trino.plugin.base.ldap.LdapClientConfig;
+import io.trino.spi.classloader.ThreadContextClassLoader;
+import io.trino.spi.security.AccessDeniedException;
+import io.trino.spi.security.GroupProvider;
+
+import javax.inject.Inject;
+import javax.naming.NamingException;
+
+import java.util.Collections;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+
+public class LdapGroupProvider
+        implements GroupProvider
+{
+    private static final Logger log = Logger.get(LdapGroupProvider.class);
+    private static final Pattern EXTRACT_GROUP_PATTERN = Pattern.compile("(?i)cn=([^,]*),(.*)$");
+    static final String GROUP_SEARCH_FILTER_PREFIX = "member=";
+
+    private final LdapCommon client;
+
+    private final Optional<String> userBaseDistinguishedName;
+    private final Optional<String> groupBaseDistinguishedName;
+    private final Optional<String> bindDistinguishedName;
+    private final Optional<String> bindPassword;
+    private final boolean allowUserNotExist;
+
+    private final LoadingCache<String, Set<String>> groupCache;
+
+    @Inject
+    public LdapGroupProvider(LdapCommon client, LdapClientConfig ldapConfig)
+    {
+        this.client = requireNonNull(client, "client is null");
+        requireNonNull(ldapConfig, "ldapConfig is null");
+
+        this.userBaseDistinguishedName = Optional.ofNullable(ldapConfig.getUserBaseDistinguishedName());
+        this.groupBaseDistinguishedName = Optional.ofNullable(ldapConfig.getGroupBaseDistinguishedName());
+        this.bindDistinguishedName = Optional.ofNullable(ldapConfig.getBindDistinguishedName());
+        this.bindPassword = Optional.ofNullable(ldapConfig.getBindPassword());
+        this.allowUserNotExist = ldapConfig.isAllowUserNotExist();
+
+        checkArgument(bindDistinguishedName.isPresent() == bindPassword.isPresent(), "Both bind distinguished name and bind password must be provided");
+        checkArgument(groupBaseDistinguishedName.isPresent(), "Base distinguished name (DN) for group must be provided");
+
+        this.groupCache = EvictableCacheBuilder.newBuilder()
+                .expireAfterWrite(ldapConfig.getLdapCacheTtl().toMillis(), MILLISECONDS)
+                .build(CacheLoader.from(this::getGroupsWithBindDistinguishedName));
+    }
+
+    @VisibleForTesting
+    void invalidateCache()
+    {
+        groupCache.invalidateAll();
+    }
+
+    @Override
+    public Set<String> getGroups(String user)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
```

Comment:
Did we see any missing class error ?

Reply:
As I told you on Slack, I'm getting errors while running the tests. I'm not 100% this will get a missing class error, but I'm looking to fix the test locally.

Reply:
Update on this: I found the problem and I can run the tests locally. I'll be working on fixing the tests today.

Reply:
Do we still see this missing class error ?